tests/job_runner_test.cc

```diff
@@ -162,20 +162,46 @@ TEST(JobRunnerTest, AlwaysActiveThread) {
         .always_active_thread_num_ = kAlwaysActiveThreadNum,
         .active_time_              = kActiveTime,
     };
-    JobRunner runner(config);
+    auto runner = JobRunner::MakeJobRunner(config);
 
     // Let the workers start.
     std::this_thread::sleep_for(std::chrono::milliseconds(100));
 
-    EXPECT_EQ(runner.ThreadNum(), kThreadNum);
-    EXPECT_EQ(runner.ActiveThreadNum(), kThreadNum);
+    EXPECT_EQ(runner->ThreadNum(), kThreadNum);
+    EXPECT_EQ(runner->ActiveThreadNum(), kThreadNum);
 
     // Within the active time, all threads are still active.
     std::this_thread::sleep_for(kActiveTime / 2);
-    EXPECT_EQ(runner.ActiveThreadNum(), kThreadNum);
+    EXPECT_EQ(runner->ActiveThreadNum(), kThreadNum);
 
     // After the active time, only "always acitve" numbeggr of threads are active.
-    EVENTUALLY_EQ(runner.ActiveThreadNum(), kAlwaysActiveThreadNum);
+    EVENTUALLY_EQ(runner->ActiveThreadNum(), kAlwaysActiveThreadNum);
+}
+
+TEST(JobRunnerTest, StrandTest) {
+    static constexpr std::size_t kThreadNum = 4;
+    static constexpr std::size_t kJobNum    = 50000;
+
+    JobRunner::Config config = {
+        .thread_num_ = kThreadNum,
+    };
+    auto runner = JobRunner::MakeJobRunner(config);
+    auto strand = runner->MakeStrand();
+
+    // Strand guarantees order, so no need to use lock/atomic variables.
+    std::size_t expected_current_job_idx = 0;
+    for (std::size_t job_idx = 0; job_idx < kJobNum; ++job_idx) {
+        EXPECT_TRUE(runner->AddJob(
+            [job_idx, &expected_current_job_idx]() { EXPECT_EQ(job_idx, expected_current_job_idx++); },
+            strand));
+    }
+
+    // Use a different atomic flag for completion instead of the counter,
+    // because TSAN might catch the data race on counter if we accidentally have concurrency on serialized jobs.
```

Comment:
Wait...that sounds like a feature.
I don't understand what's happening here, but seems you can rely on tsan to flag that "accidental concurrency"?

Reply:
Yes. The counter is not thread-safe at all so TSAN will catch it if there is unexpected concurrency.

Reply:
Then can we have our job creating data race intentionally to avoid unexpected concurrency?